doc/specs/#140 - ZIP Support.md

```diff
@@ -0,0 +1,72 @@
+- Single or multiple portable exes bundled as a suite inside an archive file.
+
+## ZIP Extraction
+The extraction of ZIPs will be done using Windows Shell APIs. ZIP files can be represented as a [ShellFolder](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ishellfolder) object that can be used to manage the contents of the ZIP file. File contents are represented as [ShellItems](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ishellitem), which can be handled using the methods exposed by the [IFileOperation interface](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ifileoperation). If the `NestedInstallerType` is non-portable, the installer file will be copied to a temporary location and used during execution of the installation flow. If the `NestedInstallerType` is portable, all files will be moved to the appropriate install location and symlinks will be created for each `NestedInstallerFile` entry.
+
+> Since we are utilizing Windows Shell APIs, we need to ensure that we are not invoking a UI during the extraction. This can be done by [setting the operation flag to not display any UI](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifileoperation-setoperationflags). 
+
+## ZIP Threat Detection
+ZIP and other archive file types are known threat vectors for malware in the form of ZIP bombs, which utilize layers of nested zip files which decompress into extremely large amounts of files. In order to protect our users and warn them before installing from any suspicious ZIP files, we will need to implement a basic archive file malware check within the client. 
+
+The check will be as follows:
+1. Recursively traverse through all files contained in an archive file.
+2. If a nested ZIP file is detected, and the number of zip layers currently extracted is less than 3, extract the contents, increment the layer count and continue traversing. Otherwise, break out of the traversal and warn the user.
+
+> ZIP headers are not always reliable when comparing compressed size vs uncompressed size as they can be modified externally. ZIP headers also differ from that of .cab or .tar.gz headers, therefore we should avoid taking a dependency on the data acquired from headers. 
+
+## Supporting Multiple Portables
+Currently, information regarding a single installed portable is stored in the ARP entry. In order to support installing a suite of portables contained inside an archive file, an additional table of information will need to be appended to the tracking catalog in order to support multiple installs under a single package. This table should replace most of the uninstall-related information that is stored in ARP for a given portable package.
+
+The table should contain the following information for each portable-related file to be installed/copied from the archive file:

```

Comment:
This is not only for multiple nested portables; it is for any archive with a nested portable and more than one file in the archive.

The table should contain a list of files that we created and their properties so that we can verify that they are the same and remove them (or not the same and leave them).

Path, flags, hash and/or secondary data (symlink target).

Path is the file system item we created.
Flags would indicate things like: directory, symlink, etc.
Hash would be for files with contents.
Symlink target for symlinks.

We only (and always) write things here *that we create*, so there would be no need to store that we created the root directory.